internal/endpoints/network.go

```diff
@@ -121,14 +126,24 @@ func (n *Network) Serve() {
 	}()
 }
 
-// Close the listener.
+// Close the listener and server.
+// Will attempt to close the server gracefully, if configured with a drain connections timeout.
+// Note that graceful shutdown will timeout if the connections do not finish (e.g.: a request caused the server
+// to Close the endpoints on the same goroutine).
 func (n *Network) Close() error {
 	if n.listener == nil {
 		return nil
 	}
 
 	logger.Info("Stopping REST API handler - closing https socket", logger.Ctx{"address": n.listener.Addr()})
-	n.cancel()
+	defer n.cancel()
+
+	// .Close() will mean that we'll no longer accept connections.
+	// It does not shutdown the server, or its currently accepted connections.
+	err := n.listener.Close()
```

Comment:
That is an interesting observation. So before Microcluster just stopped the listener but the server (and active connections) where killed as soon as the main routine exited out?

From what I saw in the docs both `server.Shutdown()` and `server.Stop()` also close the listeners so isn't it enough to just go with what you have proposed in `shutdownServer`?

Reply:
Correct, which is why we saw the `EOF` errors, but those `EOF` errors mentioned were non-deterministic simply because it was a race between finishing up the request and writing the response and the main routine exiting out. If we forcefully close a connection before the reply is written, the client would get an `EOF` error.

And yeah, those operations will also close the listeners from the start, so we can just go with `shutdownServer`.

Reply:
It turns out we **do** need to close the listener separately, as it is not passed to the server if `.Listen()` is not called, which would result in a leaked listener.

Reply:
Ah so you mean if we don't call `endpoint.Serve()` the listener isn't tracked by the server so a `server.Stop()` wouldn't close the listeners that we have created separately?

Reply:
Correct.